core/types/types.go

```diff
@@ -73,10 +72,27 @@ func NewBlockHeader(parentHash common.Hash, number *big.Int, stateRoot common.Ha
 	return bh, nil
 }
 
+// SafeCopyHeader returns a deep copy of the header to prevent side effects down the road
+func SafeCopyHeader(header *Header) *Header {
+	//copy everything but pointers / array
+	safeCopyHeader := *header
+	//copy number ptr
+	if header.Number != nil {
+		safeCopyHeader.Number = new(big.Int).Set(header.Number)
+	}
+	//copy digest byte array
+	if len(header.Digest) > 0 {
+		safeCopyHeader.Digest = make([][]byte, len(header.Digest))
+		copy(safeCopyHeader.Digest, header.Digest)
+	}
+
+	return &safeCopyHeader
+}
+
```

Comment:
I think a block builder pattern could avoid things like this. If we treat every block/header as immutable by using getters for all the fields instead of exporting them, we can limit allocations to only occur when we fetch from the DB. Please let me know your thoughts.

Reply:
Block and Header will be used also by the RPC and perhaps other instances. Doing a safe copy will avoid struggles as I can see looking to other implementations. We can always remove it later if it turns out to be useless.

Reply:
also, maybe change this to a receiver function? `func (h *Header) DeepCopy() *Header`

Reply:
Why changing it into a receiver going to be any better than it is now apart from your opinion ?

Reply:
if it is a receiver, we don't need to import `types` in other code that might use the Header type and want to copy it, we can just do `header.DeepCopy()`

Reply:
Well, if someone is going to call this method, its already importing types, otherwise the parameter of the function could not be instantiated/passed. 
I see your point and it makes sense, I can refactor.

Reply:
updated, please re-check.